content/documentation/pages/6-recipes/7-scaling/2-autoscaling.md

```diff
@@ -0,0 +1,198 @@
+if rateDifference == 0 for 3 minutes do fire ZeroThroughputDifference // <3>
+```
+
+- <1> Query expression that computes, in real-time, the throughput (e.g. rate) difference between the `time` and `transform` applications.
+- <2> `HighThroughputDifference` alert rule that fires when the rate difference exceed `500 msg/s` for the duration of `1 min`.
+- <3> `ZeroThroughputDifference` alert rule that fires if the rate difference stays `0 msg/s` for at least `3 min`.
+
+Following diagram shows the high level architecture.
+
+![SCDF autoscaling architecture](images/scdf-auto-scale-out-architecture.png)
+
+The [Data Flow metrics architecture](%currentPath%/concepts/monitoring/) is designed with the help of the Micrometer library. Prometheus can be one of the monitoring backends that can collect various application metrics for performance analysis, and it allows for alert configurations, as well.
+
+The alerting in Prometheus is divided into:
+
+- `Alert Rules` - defined inside and triggered by the Prometheus service.
+- `Alertmanager` - a standalone service that receives and manages the fired alerts and in turn sends out notifications to a pre-registered webhooks.
+
+The [Alert Rules](https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/) are based on the [Prometheus Expression Language (PQL)](https://prometheus.io/docs/prometheus/latest/querying/basics/).
+They are used to define and send the scale alerts to `Alertmanager`.
+For example the scale-out alert rule definition looks something like this:
+
+```sql
+alert: HighThroughputDifference
+expr:  avg(irate(spring_integration_send_seconds_count{application_name="time"}[1m])) by(stream_name) -
+       avg(irate(spring_integration_send_seconds_count{application_name="transform"}[1m])) by(stream_name) > 500
+for: 30s
+```
+
+<!--TIP-->
+
+Check [alert.rule.yml](TODO) (and [here](TODO) for the kubectl installation) to see the used alert rules definitions.
+
+<!--END_TIP-->
+
+<!--TIP-->
+
+The `spring_integration_send_seconds_count` metrics comes from the `spring integration` [micrometer support](https://docs.spring.io/spring-integration/docs/5.2.2.RELEASE/reference/html/system-management.html#micrometer-integration) and is used for computing the message rate.
+
+<!--END_TIP-->
+
+The [Alertmanager](https://prometheus.io/docs/alerting/alertmanager) is a standalone service that manages the alerts, including silencing, inhibition, aggregation and sending out notifications to pre-configured webhooks.
+
+The [AlertWebHookApplication](TODO) is a custom Spring Boot application, registered as a [Alertmanager Webhook Receiver](https://prometheus.io/docs/alerting/configuration/#webhook_config) via the [config.yml](TODO).
+The `AlertWebHookApplication` receives the alert notifications (in JSON format) and with the help of SCDF's Scale API triggers the scale-out request to autoscale the given streaming data pipelines.
```

Comment:
```suggestion
The `AlertWebHookApplication` receives the alert notifications (in JSON format) from Prometheus. With the help of SCDF's Scale API, it can then trigger the scale-out request to autoscale the given streaming data pipelines in SCDF.
```